File: python/open_guardrail/guards/api_response_contract.py
"""Validate structured API responses match expected contract."""
from __future__ import annotations

import json
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_SENSITIVE = ["password", "secret", "token", "api_key", "apikey", "private_key"]
_MAX_LEN = 10000


class _ApiResponseContract:
    def __init__(self, *, action: str = "warn", required_fields: Optional[List[str]] = None) -> None:
        self.name = "api-response-contract"
        self.action = action
        self.required_fields = required_fields or []

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        try:
            parsed = json.loads(text.strip())
        except (json.JSONDecodeError, ValueError):
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="api-response-contract",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        if not isinstance(parsed, dict):
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="api-response-contract",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        issues: list[str] = []
        if not any(k in parsed for k in ("status", "data", "result")):
            issues.append("missing-status-or-data-field")
        for f in self.required_fields:
            if f not in parsed:
                issues.append(f"missing-required:{f}")
            elif parsed[f] is None:
                issues.append(f"null-required:{f}")
        for key, val in parsed.items():
            if any(s in key.lower() for s in _SENSITIVE):
                issues.append(f"sensitive-key:{key}")
            if isinstance(val, str) and len(val) > _MAX_LEN:
                issues.append(f"oversized-field:{key}")
        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="api-response-contract",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"API contract issues: {', '.join(issues)}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"issues": issues} if triggered else None,
        )


def api_response_contract(*, action: str = "warn", required_fields: Optional[List[str]] = None) -> _ApiResponseContract:
    return _ApiResponseContract(action=action, required_fields=required_fields)

File: packages/guards/src/api-response-contract.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface ApiResponseContractOptions {
  action: 'block' | 'warn';
  requiredFields?: string[];
}

const SENSITIVE_KEYS = ['password', 'secret', 'token', 'api_key', 'apikey', 'private_key'];
const MAX_FIELD_LENGTH = 10000;

function checkObject(obj: Record<string, unknown>, requiredFields: string[]): string[] {
  const issues: string[] = [];

  // Check for status/data/result presence
  const hasStructure = 'status' in obj || 'data' in obj || 'result' in obj;
  if (!hasStructure) issues.push('missing-status-or-data-field');

  // Required fields check
  for (const f of requiredFields) {
    if (!(f in obj)) {
      issues.push(`missing-required:${f}`);
    } else if (obj[f] === null || obj[f] === undefined) {
      issues.push(`null-required:${f}`);
    }
  }

  // Check for sensitive keys and field size
  for (const [key, val] of Object.entries(obj)) {
    if (SENSITIVE_KEYS.some((s) => key.toLowerCase().includes(s))) {
      issues.push(`sensitive-key:${key}`);
    }
    if (typeof val === 'string' && val.length > MAX_FIELD_LENGTH) {
      issues.push(`oversized-field:${key}`);
    }
  }

  return issues;
}

export function apiResponseContract(options: ApiResponseContractOptions): Guard {
  return {
    name: 'api-response-contract',
    version: '0.1.0',
    description: 'Validate structured API responses match expected contract',
    category: 'format',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      let parsed: unknown;
      try {
        parsed = JSON.parse(text.trim());
      } catch {
        // Not JSON — guard does not apply
        return {
          guardName: 'api-response-contract',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
        return {
          guardName: 'api-response-contract',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      const issues = checkObject(parsed as Record<string, unknown>, options.requiredFields ?? []);
      const triggered = issues.length > 0;
      return {
        guardName: 'api-response-contract',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered ? `API contract issues: ${issues.join(', ')}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
